common/config/src/main/java/io/helidon/common/config/Config.java

```diff
@@ -0,0 +1,346 @@
+     *
+     * @return {@code true} if the node is existing leaf node, {@code false}
+     *         otherwise.
+     */
+    boolean isLeaf();
+
+    /**
+     * Returns {@code true} if this node exists and is Type#Object.
+     *
+     * @return {@code true} if the node exists and is Type#Object, {@code false}
+     *         otherwise.
+     */
+    boolean isObject();
+
+    /**
+     * Returns {@code true} if this node exists and is Type#List.
+     *
+     * @return {@code true} if the node exists and is Type#List, {@code false}
+     *         otherwise.
+     */
+    boolean isList();
+
+    //
+    // accessors
+    //
+
+    /**
+     * Typed value as a {@link ConfigValue}.
+     *
+     * @param type type class
+     * @param <T>  type
+     * @return typed value
+     * @see ConfigValue#map(java.util.function.Function)
+     * @see ConfigValue#get()
+     * @see ConfigValue#orElse(Object)
+     */
+    <T> ConfigValue<T> as(Class<T> type);
+
+    // shortcut methods
+
+    /**
+     * Returns list of specified type.
+     *
+     * @param type type class
+     * @param <T>  type of list elements
+     * @return a typed list with values
+     * @throws io.helidon.common.config.ConfigException in case of problem to map property value.
+     */
+    default <T> ConfigValue<List<T>> asList(Class<T> type) throws ConfigException {
+        throw new ConfigException("unsupported; do you have a valid config module in the classpath?");
+    }
+
+    /**
+     * Transform all leaf nodes (values) into Map instance.
+     *
+     * @return new Map instance that contains all config leaf node values
+     * @throws io.helidon.common.config.ConfigException in case the node is Type#MISSING.
+     */
+    default ConfigValue<Map<String, String>> asMap() throws ConfigException {
+        throw new ConfigException("unsupported; do you have a valid config module in the classpath?");
+    }
+
+    /**
+     * Object represents fully-qualified key of config node.
+     * <p>
+     * Fully-qualified key is list of key tokens separated by {@code .} (dot character).
+     * Depending on context the key token is evaluated one by one:
+     * <ul>
+     * <li>in Type#OBJECT node the token represents a <strong>name of object member</strong>;</li>
+     * <li>in Type#LIST node the token represents an zero-based <strong>index of list
+     * element</strong>,
+     * an unsigned base-10 integer value, leading zeros are not allowed.</li>
+     * </ul>
+     * <p>
+     * The ABNF syntax of config key is:
+     * <pre>{@code
+     * config-key = *1( key-token *( "." key-token ) )
+     *  key-token = *( unescaped / escaped )
+     *  unescaped = %x00-2D / %x2F-7D / %x7F-10FFFF
+     *            ; %x2E ('.') and %x7E ('~') are excluded from 'unescaped'
+     *    escaped = "~" ( "0" / "1" )
+     *            ; representing '~' and '.', respectively
+     * }</pre>
+     *
+     * @see Config#key()
+     */
+    interface BaseKey extends Comparable<BaseKey> {
```

Comment:
This should be named `Key`. BaseKey would mean an abstract class. Also this is one of the main APIs of Config, so not prefixes should be used.